File: Java/src/main/java/com/mexc/example/futures/api/MexcMarketDataApi.java
package com.mexc.example.futures.api;

import com.google.gson.JsonObject;
import com.google.gson.JsonArray;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * MEXC Futures Market Data API
 * <p>
 * Implements all public market data endpoints
 * Inherits from MexcOkHttpClient base class
 * <p>
 * All endpoints are public and do not require authentication.
 */
public class MexcMarketDataApi extends MexcOkHttpClient {

    public MexcMarketDataApi() {
        super(); // Public client - no API keys needed
    }

    /**
     * Get Server Time (Ping)
     * GET /api/v1/contract/ping
     * <p>
     * Test connectivity and get server time
     */
    public JsonObject ping() throws IOException {
        System.out.println("\n===  Ping - Get Server Time ===");
        JsonObject response = get("/api/v1/contract/ping", null);
        if (response != null && response.get("success").getAsBoolean()) {
            System.out.println("Server timestamp: " + response.get("data"));
        }
        return response;
    }

    /**
     * Get Contract Information
     * GET /api/v1/contract/detail
     *
     * @param symbol Optional contract symbol (e.g., "BTC_USDT"). Returns all if null.
     */
    public JsonObject getContractDetail(String symbol) throws IOException {
        System.out.println("\n===  Get Contract Information" +
                (symbol != null ? ": " + symbol : " (All)") + " ===");

        Map<String, String> params = null;
        if (symbol != null) {
            params = new HashMap<>();
            params.put("symbol", symbol);
        }

        JsonObject response = get("/api/v1/contract/detail", params);

        if (response != null && response.get("success").getAsBoolean()) {
            if (symbol != null) {
                JsonObject data = response.getAsJsonObject("data");
                System.out.println("Symbol: " + data.get("symbol"));
                System.out.println("Base Coin: " + data.get("baseCoin"));
                System.out.println("Quote Coin: " + data.get("quoteCoin"));
                System.out.println("Contract Size: " + data.get("contractSize"));
                System.out.println("Max Leverage: " + data.get("maxLeverage"));
            } else {
                JsonArray data = response.getAsJsonArray("data");
                System.out.println("Retrieved " + data.size() + " contracts");
            }
        }
        return response;
    }

    /**
     * Get Transferable Currencies
     * GET /api/v1/contract/support_currencies
     */
    public JsonObject getSupportCurrencies() throws IOException {
        System.out.println("\n===  Get Transferable Currencies ===");
        JsonObject response = get("/api/v1/contract/support_currencies", null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonArray currencies = response.getAsJsonArray("data");
            System.out.println("Supported currencies: " + currencies.size());
        }
        return response;
    }

    /**
     * Get Order Book Depth
     * GET /api/v1/contract/depth/{symbol}
     *
     * @param symbol Contract symbol (required)
     * @param limit  Number of depth levels (optional, default 100)
     */
    public JsonObject getDepth(String symbol, Integer limit) throws IOException {
        System.out.println("\n===  Get Order Book Depth: " + symbol + " ===");

        String endpoint = "/api/v1/contract/depth/" + symbol;
        Map<String, String> params = null;
        if (limit != null) {
            params = new HashMap<>();
            params.put("limit", limit.toString());
        }

        JsonObject response = get(endpoint, params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Order Book Version: " + data.get("version"));
            System.out.println("Timestamp: " + data.get("timestamp"));
            System.out.println("Bids: " + data.getAsJsonArray("bids").size() + " levels");
            System.out.println("Asks: " + data.getAsJsonArray("asks").size() + " levels");
        }
        return response;
    }

    /**
     * Get Last N Depth Snapshots
     * GET /api/v1/contract/depth_commits/{symbol}/{limit}
     *
     * @param symbol Contract symbol (required)
     * @param limit  Number of snapshots to return (required)
     */
    public JsonObject getDepthCommits(String symbol, int limit) throws IOException {
        System.out.println("\n===  Get Last " + limit + " Depth Snapshots: " + symbol + " ===");

        String endpoint = "/api/v1/contract/depth_commits/" + symbol + "/" + limit;
        JsonObject response = get(endpoint, null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonArray snapshots = response.getAsJsonArray("data");
            System.out.println("Retrieved " + snapshots.size() + " snapshots");
        }
        return response;
    }

    /**
     * Get Index Price
     * GET /api/v1/contract/index_price/{symbol}
     *
     * @param symbol Contract symbol (required)
     */
    public JsonObject getIndexPrice(String symbol) throws IOException {
        System.out.println("\n=== Get Index Price: " + symbol + " ===");

        String endpoint = "/api/v1/contract/index_price/" + symbol;
        JsonObject response = get(endpoint, null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Index Price: " + data.get("indexPrice"));
        }
        return response;
    }

    /**
     * Get Fair Price
     * GET /api/v1/contract/fair_price/{symbol}
     *
     * @param symbol Contract symbol (required)
     */
    public JsonObject getFairPrice(String symbol) throws IOException {
        System.out.println("\n===  Get Fair Price: " + symbol + " ===");

        String endpoint = "/api/v1/contract/fair_price/" + symbol;
        JsonObject response = get(endpoint, null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Fair Price: " + data.get("fairPrice"));
        }
        return response;
    }

    /**
     * Get Funding Rate
     * GET /api/v1/contract/funding_rate/{symbol}
     *
     * @param symbol Contract symbol (required)
     */
    public JsonObject getFundingRate(String symbol) throws IOException {
        System.out.println("\n=== Get Funding Rate: " + symbol + " ===");

        String endpoint = "/api/v1/contract/funding_rate/" + symbol;
        JsonObject response = get(endpoint, null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Funding Rate: " + data.get("fundingRate"));
            System.out.println("Next Settlement: " + data.get("nextSettleTime"));
        }
        return response;
    }

    /**
     * Get Kline/Candlestick Data
     * GET /api/v1/contract/kline/{symbol}
     *
     * @param symbol   Contract symbol (required)
     * @param interval Kline interval: 1m,5m,15m,30m,1h,4h,1d,1w,1M
     * @param start    Optional start time in milliseconds
     * @param end      Optional end time in milliseconds
     * @param limit    Optional number of records (default 100, max 500)
     */
    public JsonObject getKline(String symbol, String interval, Long start, Long end, Integer limit) throws IOException {
        System.out.println("\n===  Get Kline Data: " + symbol + " " + interval + " ===");

        String endpoint = "/api/v1/contract/kline/" + symbol;
        Map<String, String> params = new HashMap<>();
        params.put("interval", interval);

        if (start != null) params.put("start", String.valueOf(start));
        if (end != null) params.put("end", String.valueOf(end));
        if (limit != null) params.put("limit", String.valueOf(limit));

        JsonObject response = get(endpoint, params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject klineData = response.getAsJsonObject("data");
            System.out.println("Retrieved " + klineData.size() + " kline records");
            if (!klineData.isEmpty()) {
                JsonArray jsonArray = klineData.getAsJsonArray("close");
                System.out.println("First: " + jsonArray.get(0));
            }
        }
        return response;
    }

    /**
     * Get Index Price Kline Data
     * GET /api/v1/contract/kline/index_price/{symbol}
     * <p>
     * Retrieves historical index price candlestick data.
     *
     * @param symbol   Contract symbol (required, e.g., "BTC_USDT")
     * @param interval Kline interval. Supported values:
     *                 Min1, Min5, Min15, Min30, Hour1, Hour4, Hour8, Day1, Week1, Month1
     * @param start    Start time in seconds (Unix timestamp, required)
     * @param end      End time in seconds (Unix timestamp, required)
     * @return JsonObject containing arrays of time, open, close, high, low, vol, amount, etc.
     */
    public JsonObject getIndexPriceKline(String symbol, String interval, long start, long end) throws IOException {
        System.out.println("\n===  Get Index Price Kline: " + symbol + " " + interval + " ===");
        System.out.println("Time range: " + start + " to " + end);

        String endpoint = "/api/v1/contract/kline/index_price/" + symbol;
        Map<String, String> params = new HashMap<>();
        params.put("interval", interval);
        params.put("start", String.valueOf(start));
        params.put("end", String.valueOf(end));

        JsonObject response = get(endpoint, params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            JsonArray timeArray = data.getAsJsonArray("time");
            System.out.println("Retrieved " + timeArray.size() + " index price kline records");
            // Optional: Print sample data
            if (!timeArray.isEmpty()) {
                System.out.println("First record - Time: " + timeArray.get(0) +
                        ", Open: " + data.getAsJsonArray("open").get(0) +
                        ", Close: " + data.getAsJsonArray("close").get(0));
            }
        }
        return response;
    }

    /**
     * Get Fair Price Kline Data
     * GET /api/v1/contract/kline/fair_price/{symbol}
     * <p>
     * Retrieves historical fair price candlestick data.
     *
     * @param symbol   Contract symbol (required, e.g., "BTC_USDT")
     * @param interval Kline interval. Supported values:
     *                 Min1, Min5, Min15, Min30, Hour1, Hour4, Hour8, Day1, Week1, Month1
     * @param start    Start time in seconds (Unix timestamp, required)
     * @param end      End time in seconds (Unix timestamp, required)
     * @return JsonObject containing arrays of time, open, close, high, low, vol, amount, etc.
     */
    public JsonObject getFairPriceKline(String symbol, String interval, long start, long end) throws IOException {
        System.out.println("\n===  Get Fair Price Kline: " + symbol + " " + interval + " ===");
        System.out.println("Time range: " + start + " to " + end);

        String endpoint = "/api/v1/contract/kline/fair_price/" + symbol;
        Map<String, String> params = new HashMap<>();
        params.put("interval", interval);
        params.put("start", String.valueOf(start));
        params.put("end", String.valueOf(end));

        JsonObject response = get(endpoint, params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            JsonArray timeArray = data.getAsJsonArray("time");
            System.out.println("Retrieved " + timeArray.size() + " fair price kline records");

            // Optional: Print sample data
            if (!timeArray.isEmpty()) {
                System.out.println("First record - Time: " + timeArray.get(0) +
                        ", Open: " + data.getAsJsonArray("open").get(0) +
                        ", Close: " + data.getAsJsonArray("close").get(0));
            }
        }
        return response;
    }

    /**
     * Get Recent Deals/Trades
     * GET /api/v1/contract/deals/{symbol}
     * <p>
     * Retrieves the most recent trades/deals for a specific symbol.
     * This is similar to recent trades but may include additional information.
     *
     * @param symbol Contract symbol (required)
     * @param limit  Number of deals to return (default 100, max 500)
     * @return JsonObject containing recent deals
     */
    public JsonObject getRecentDeals(String symbol, Integer limit) throws IOException {
        System.out.println("\n=== Get Recent Deals: " + symbol + " ===");

        String endpoint = "/api/v1/contract/deals/" + symbol;
        Map<String, String> params = null;
        if (limit != null) {
            params = new HashMap<>();
            params.put("limit", limit.toString());
        }

        JsonObject response = get(endpoint, params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonArray data = response.getAsJsonArray("data");
            System.out.println("Retrieved " + data.size() + " recent deals");

            if (!data.isEmpty()) {
                System.out.println("\nMost recent deals:");
                int count = Math.min(5, data.size());
                for (int i = 0; i < count; i++) {
                    JsonObject deal = data.get(i).getAsJsonObject();
                    // Format depends on API response - adjust as needed
                    System.out.println("  " + (i + 1) + ": " + deal);
                }

                // Calculate some statistics
                double totalVolume = 0;
                double totalAmount = 0;
                for (int i = 0; i < data.size(); i++) {
                    JsonObject deal = data.get(i).getAsJsonObject();
                    if (deal.size() >= 2) {
                        totalVolume += deal.get("v").getAsDouble();
                        totalAmount += deal.get("p").getAsDouble() * deal.get("v").getAsDouble();
                    }
                }
                System.out.printf("\nSummary - Total Volume: %.4f, Total Amount: %.2f%n",
                        totalVolume, totalAmount);
            }
        }
        return response;
    }

    /**
     * 23. Get All Tickers
     * GET /api/v1/contract/ticker
     * <p>
     * Retrieves all ticker information in a single call.
     * This is a comprehensive endpoint that returns price, volume, change, etc. for all symbols.
     *
     * @return JsonObject containing all ticker data
     */
    public JsonObject getAllTickers() throws IOException {
        System.out.println("\n===  Get All Tickers ===");

        JsonObject response = get("/api/v1/contract/ticker", null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonArray data = response.getAsJsonArray("data");
            System.out.println("Retrieved tickers for " + data.size() + " symbols");

            // Print summary for major symbols
            String[] majorSymbols = {"BTC_USDT", "ETH_USDT", "BNB_USDT"};
            for (String symbol : majorSymbols) {
                for (int i = 0; i < data.size(); i++) {
                    JsonObject ticker = data.get(i).getAsJsonObject();
                    if (symbol.equals(ticker.get("symbol").getAsString())) {
                        System.out.println("\n" + symbol + ":");
                        System.out.println("  Last: " + ticker.get("lastPrice"));
                        System.out.println("  riseFallRate: " + ticker.get("riseFallValue") +
                                " (" + ticker.get("riseFallRate") + "%)");
                        System.out.println("  Volume: " + ticker.get("volume24"));
                        break;
                    }
                }
            }
        }
        return response;
    }

    /**
     * Get Risk Reverse by Symbol
     * GET /api/v1/contract/risk_reverse/{symbol}
     * <p>
     * Retrieves current risk reverse (insurance fund) information for a specific symbol.
     *
     * @param symbol Contract symbol (required)
     * @return JsonObject containing risk reverse information
     */
    public JsonObject getRiskReverseBySymbol(String symbol) throws IOException {
        System.out.println("\n===  Get Risk Reverse: " + symbol + " ===");

        String endpoint = "/api/v1/contract/risk_reverse/" + symbol;
        JsonObject response = get(endpoint, null);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Symbol: " + data.get("symbol"));
            System.out.println("Risk Reverse Amount: " + data.get("available"));
            System.out.println("Update Time: " + data.get("timestamp"));
        }
        return response;
    }


    /**
     * Get Risk Reverse History
     * GET /api/v1/contract/risk_reverse/history
     * <p>
     * Retrieves historical risk reverse (insurance fund) data.
     *
     * @param symbol   Optional contract symbol
     * @param pageNum  Page number (default 1)
     * @param pageSize Page size (default 20, max 100)
     * @return JsonObject containing risk reverse history
     */
    public JsonObject getRiskReverseHistory(String symbol, Integer pageNum, Integer pageSize) throws IOException {
        System.out.println("\n===  Get Risk Reverse History" + (symbol != null ? ": " + symbol : "") + " ===");

        Map<String, String> params = new HashMap<>();
        if (symbol != null) params.put("symbol", symbol);
        if (pageNum != null) params.put("page_num", String.valueOf(pageNum));
        if (pageSize != null) params.put("page_size", String.valueOf(pageSize));

        JsonObject response = get("/api/v1/contract/risk_reverse/history", params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Total records: " + data.get("totalCount"));
            System.out.println("Current page: " + data.get("currentPage"));
            System.out.println("Page size: " + data.get("pageSize"));

            JsonArray resultList = data.getAsJsonArray("resultList");
            System.out.println("Records in this page: " + resultList.size());

            if (!resultList.isEmpty()) {
                JsonObject first = resultList.get(0).getAsJsonObject();
                System.out.println("Sample - Symbol: " + first.get("symbol") +
                        ", Amount: " + first.get("amount") +
                        ", Time: " + first.get("createTime"));
            }
        }
        return response;
    }

    /**
     * Get Funding Rate History
     * GET /api/v1/contract/funding_rate/history
     * <p>
     * Retrieves historical funding rate data for all symbols or specific symbol.
     *
     * @param symbol Optional contract symbol (e.g., "BTC_USDT"). Returns all if null.
     * @param limit  Number of records (default 100, max 500)
     * @return JsonObject containing funding rate history
     */
    public JsonObject getFundingRateHistory(String symbol, Integer limit) throws IOException {
        System.out.println("\n===  Get Funding Rate History" + (symbol != null ? ": " + symbol : " (All)") + " ===");

        Map<String, String> params = new HashMap<>();
        if (symbol != null) params.put("symbol", symbol);
        if (limit != null) params.put("limit", String.valueOf(limit));

        JsonObject response = get("/api/v1/contract/funding_rate/history", params);

        if (response != null && response.get("success").getAsBoolean()) {
            JsonObject data = response.getAsJsonObject("data");
            System.out.println("Retrieved " + data.size() + " funding rate records");

            if (!data.isEmpty()) {
                JsonArray array = data.get("resultList").getAsJsonArray();
                JsonObject first = array.get(0).getAsJsonObject();
                System.out.println("Sample - Symbol: " + first.get("symbol") +
                        ", Rate: " + first.get("fundingRate") +
                        ", Time: " + first.get("settleTime"));
            }
        }
        return response;
    }
}

File: dotnet/futures/Mexc.Client/MexcMarketDataApi.cs
using System.Text.Json;
namespace Mexc.Client
{
    /// <summary>
    /// MEXC Futures Market Data API
    /// All endpoints are public and do not require authentication.
    /// </summary>
    public class MexcMarketDataApi : MexcHttpClient
    {
        public MexcMarketDataApi(ILogger logger = null) : base(null, null, logger)
        {
        }

        /// <summary>
        /// Get Server Time (Ping)
        /// GET /api/v1/contract/ping
        /// </summary>
        public async Task<JsonDocument> PingAsync()
        {
            _logger?.LogInformation("\n=== Ping - Get Server Time ===");
            var response = await GetAsync("/api/v1/contract/ping");
            
            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    _logger?.LogInformation($"Server timestamp: {data}");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Contract Information
        /// GET /api/v1/contract/detail
        /// </summary>
        public async Task<JsonDocument> GetContractDetailAsync(string symbol = null)
        {
            _logger?.LogInformation($"\n=== Get Contract Information {(symbol ?? "(All)")} ===");

            var paramsDict = new Dictionary<string, string>();
            if (!string.IsNullOrEmpty(symbol))
            {
                paramsDict["symbol"] = symbol;
            }

            var response = await GetAsync("/api/v1/contract/detail", paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (!string.IsNullOrEmpty(symbol))
                {
                    if (response.RootElement.TryGetProperty("data", out var data))
                    {
                        _logger?.LogInformation($"Symbol: {data.GetProperty("symbol")}");
                        _logger?.LogInformation($"Base Coin: {data.GetProperty("baseCoin")}");
                        _logger?.LogInformation($"Quote Coin: {data.GetProperty("quoteCoin")}");
                        _logger?.LogInformation($"Contract Size: {data.GetProperty("contractSize")}");
                        _logger?.LogInformation($"Max Leverage: {data.GetProperty("maxLeverage")}");
                    }
                }
                else
                {
                    if (response.RootElement.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
                    {
                        _logger?.LogInformation($"Retrieved {data.GetArrayLength()} contracts");
                    }
                }
            }
            return response;
        }

        /// <summary>
        /// Get Transferable Currencies
        /// GET /api/v1/contract/support_currencies
        /// </summary>
        public async Task<JsonDocument> GetSupportCurrenciesAsync()
        {
            _logger?.LogInformation("\n=== Get Transferable Currencies ===");
            var response = await GetAsync("/api/v1/contract/support_currencies");

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
                {
                    _logger?.LogInformation($"Supported currencies: {data.GetArrayLength()}");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Order Book Depth
        /// GET /api/v1/contract/depth/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetDepthAsync(string symbol, int? limit = null)
        {
            _logger?.LogInformation($"\n=== Get Order Book Depth: {symbol} ===");

            var endpoint = $"/api/v1/contract/depth/{symbol}";
            var paramsDict = new Dictionary<string, string>();
            if (limit.HasValue)
            {
                paramsDict["limit"] = limit.ToString();
            }

            var response = await GetAsync(endpoint, paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("version", out var version))
                        _logger?.LogInformation($"Order Book Version: {version}");
                    
                    if (data.TryGetProperty("timestamp", out var timestamp))
                        _logger?.LogInformation($"Timestamp: {timestamp}");
                    
                    if (data.TryGetProperty("bids", out var bids) && bids.ValueKind == JsonValueKind.Array)
                        _logger?.LogInformation($"Bids: {bids.GetArrayLength()} levels");
                    
                    if (data.TryGetProperty("asks", out var asks) && asks.ValueKind == JsonValueKind.Array)
                        _logger?.LogInformation($"Asks: {asks.GetArrayLength()} levels");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Last N Depth Snapshots
        /// GET /api/v1/contract/depth_commits/{symbol}/{limit}
        /// </summary>
        public async Task<JsonDocument> GetDepthCommitsAsync(string symbol, int limit)
        {
            _logger?.LogInformation($"\n=== Get Last {limit} Depth Snapshots: {symbol} ===");

            var endpoint = $"/api/v1/contract/depth_commits/{symbol}/{limit}";
            var response = await GetAsync(endpoint);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
                {
                    _logger?.LogInformation($"Retrieved {data.GetArrayLength()} snapshots");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Index Price
        /// GET /api/v1/contract/index_price/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetIndexPriceAsync(string symbol)
        {
            _logger?.LogInformation($"\n=== Get Index Price: {symbol} ===");

            var endpoint = $"/api/v1/contract/index_price/{symbol}";
            var response = await GetAsync(endpoint);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("indexPrice", out var indexPrice))
                        _logger?.LogInformation($"Index Price: {indexPrice}");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Fair Price
        /// GET /api/v1/contract/fair_price/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetFairPriceAsync(string symbol)
        {
            _logger?.LogInformation($"\n=== Get Fair Price: {symbol} ===");

            var endpoint = $"/api/v1/contract/fair_price/{symbol}";
            var response = await GetAsync(endpoint);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("fairPrice", out var fairPrice))
                        _logger?.LogInformation($"Fair Price: {fairPrice}");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Funding Rate
        /// GET /api/v1/contract/funding_rate/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetFundingRateAsync(string symbol)
        {
            _logger?.LogInformation($"\n=== Get Funding Rate: {symbol} ===");

            var endpoint = $"/api/v1/contract/funding_rate/{symbol}";
            var response = await GetAsync(endpoint);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("fundingRate", out var fundingRate))
                        _logger?.LogInformation($"Funding Rate: {fundingRate}");
                    
                    if (data.TryGetProperty("nextSettleTime", out var nextSettleTime))
                        _logger?.LogInformation($"Next Settlement: {nextSettleTime}");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Kline/Candlestick Data
        /// GET /api/v1/contract/kline/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetKlineAsync(string symbol, string interval, long? start = null, long? end = null, int? limit = null)
        {
            _logger?.LogInformation($"\n=== Get Kline Data: {symbol} {interval} ===");

            var endpoint = $"/api/v1/contract/kline/{symbol}";
            var paramsDict = new Dictionary<string, string> { { "interval", interval } };

            if (start.HasValue) paramsDict["start"] = start.ToString();
            if (end.HasValue) paramsDict["end"] = end.ToString();
            if (limit.HasValue) paramsDict["limit"] = limit.ToString();

            var response = await GetAsync(endpoint, paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var klineData))
                {
                    _logger?.LogInformation($"Retrieved kline records");
                    if (klineData.TryGetProperty("close", out var closeArray) && closeArray.ValueKind == JsonValueKind.Array)
                    {
                        if (closeArray.GetArrayLength() > 0)
                            _logger?.LogInformation($"First close: {closeArray[0]}");
                    }
                }
            }
            return response;
        }

        /// <summary>
        /// Get Index Price Kline Data
        /// GET /api/v1/contract/kline/index_price/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetIndexPriceKlineAsync(string symbol, string interval, long start, long end)
        {
            _logger?.LogInformation($"\n=== Get Index Price Kline: {symbol} {interval} ===");
            _logger?.LogInformation($"Time range: {start} to {end}");

            var endpoint = $"/api/v1/contract/kline/index_price/{symbol}";
            var paramsDict = new Dictionary<string, string>
            {
                { "interval", interval },
                { "start", start.ToString() },
                { "end", end.ToString() }
            };

            var response = await GetAsync(endpoint, paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("time", out var timeArray) && timeArray.ValueKind == JsonValueKind.Array)
                    {
                        _logger?.LogInformation($"Retrieved {timeArray.GetArrayLength()} index price kline records");
                    }
                }
            }
            return response;
        }

        /// <summary>
        /// Get Fair Price Kline Data
        /// GET /api/v1/contract/kline/fair_price/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetFairPriceKlineAsync(string symbol, string interval, long start, long end)
        {
            _logger?.LogInformation($"\n=== Get Fair Price Kline: {symbol} {interval} ===");
            _logger?.LogInformation($"Time range: {start} to {end}");

            var endpoint = $"/api/v1/contract/kline/fair_price/{symbol}";
            var paramsDict = new Dictionary<string, string>
            {
                { "interval", interval },
                { "start", start.ToString() },
                { "end", end.ToString() }
            };

            var response = await GetAsync(endpoint, paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("time", out var timeArray) && timeArray.ValueKind == JsonValueKind.Array)
                    {
                        _logger?.LogInformation($"Retrieved {timeArray.GetArrayLength()} fair price kline records");
                    }
                }
            }
            return response;
        }

        /// <summary>
        /// Get Recent Deals/Trades
        /// GET /api/v1/contract/deals/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetRecentDealsAsync(string symbol, int? limit = null)
        {
            _logger?.LogInformation($"\n=== Get Recent Deals: {symbol} ===");

            var endpoint = $"/api/v1/contract/deals/{symbol}";
            var paramsDict = new Dictionary<string, string>();
            if (limit.HasValue)
            {
                paramsDict["limit"] = limit.ToString();
            }

            var response = await GetAsync(endpoint, paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
                {
                    _logger?.LogInformation($"Retrieved {data.GetArrayLength()} recent deals");
                }
            }
            return response;
        }

        /// <summary>
        /// Get All Tickers
        /// GET /api/v1/contract/ticker
        /// </summary>
        public async Task<JsonDocument> GetAllTickersAsync()
        {
            _logger?.LogInformation("\n=== Get All Tickers ===");

            var response = await GetAsync("/api/v1/contract/ticker");

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data) && data.ValueKind == JsonValueKind.Array)
                {
                    _logger?.LogInformation($"Retrieved tickers for {data.GetArrayLength()} symbols");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Risk Reverse by Symbol
        /// GET /api/v1/contract/risk_reverse/{symbol}
        /// </summary>
        public async Task<JsonDocument> GetRiskReverseBySymbolAsync(string symbol)
        {
            _logger?.LogInformation($"\n=== Get Risk Reverse: {symbol} ===");

            var endpoint = $"/api/v1/contract/risk_reverse/{symbol}";
            var response = await GetAsync(endpoint);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("symbol", out var sym))
                        _logger?.LogInformation($"Symbol: {sym}");
                    
                    if (data.TryGetProperty("available", out var available))
                        _logger?.LogInformation($"Risk Reverse Amount: {available}");
                    
                    if (data.TryGetProperty("timestamp", out var timestamp))
                        _logger?.LogInformation($"Update Time: {timestamp}");
                }
            }
            return response;
        }

        /// <summary>
        /// Get Risk Reverse History
        /// GET /api/v1/contract/risk_reverse/history
        /// </summary>
        public async Task<JsonDocument> GetRiskReverseHistoryAsync(string symbol = null, int pageNum = 1, int pageSize = 20)
        {
            _logger?.LogInformation($"\n=== Get Risk Reverse History {(symbol ?? "")} ===");

            var paramsDict = new Dictionary<string, string>();
            if (!string.IsNullOrEmpty(symbol))
            {
                paramsDict["symbol"] = symbol;
            }
            paramsDict["page_num"] = pageNum.ToString();
            paramsDict["page_size"] = Math.Min(pageSize, 100).ToString();

            var response = await GetAsync("/api/v1/contract/risk_reverse/history", paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("totalCount", out var totalCount))
                        _logger?.LogInformation($"Total records: {totalCount}");
                    
                    if (data.TryGetProperty("currentPage", out var currentPage))
                        _logger?.LogInformation($"Current page: {currentPage}");
                    
                    if (data.TryGetProperty("pageSize", out var pageSizeProp))
                        _logger?.LogInformation($"Page size: {pageSizeProp}");

                    if (data.TryGetProperty("resultList", out var resultList) && resultList.ValueKind == JsonValueKind.Array)
                    {
                        _logger?.LogInformation($"Records in this page: {resultList.GetArrayLength()}");
                    }
                }
            }
            return response;
        }

        /// <summary>
        /// Get Funding Rate History
        /// GET /api/v1/contract/funding_rate/history
        /// </summary>
        public async Task<JsonDocument> GetFundingRateHistoryAsync(string symbol = null, int? limit = null)
        {
            _logger?.LogInformation($"\n=== Get Funding Rate History {(symbol ?? "(All)")} ===");

            var paramsDict = new Dictionary<string, string>();
            if (!string.IsNullOrEmpty(symbol))
            {
                paramsDict["symbol"] = symbol;
            }
            if (limit.HasValue)
            {
                paramsDict["limit"] = limit.ToString();
            }

            var response = await GetAsync("/api/v1/contract/funding_rate/history", paramsDict);

            if (response != null && response.RootElement.TryGetProperty("success", out var success) && success.GetBoolean())
            {
                if (response.RootElement.TryGetProperty("data", out var data))
                {
                    if (data.TryGetProperty("resultList", out var resultList) && resultList.ValueKind == JsonValueKind.Array)
                    {
                        _logger?.LogInformation($"Retrieved {resultList.GetArrayLength()} funding rate records");
                    }
                }
            }
            return response;
        }
    }
}